### PR TITLE
Enable bloom filter for _id field in tsdb indices.

### PR DESCRIPTION
**Pull request by martijnvg**

### CODE REVIEW

**Comment by martijnvg** [comment]
The following indexing performance improvements were observer running the tsdb Rally track during the index task:

```
|                                                Min Throughput |                   index | 35621.2         | 53792.8         | 18171.6     | docs/s |  +51.01% |
|                                               Mean Throughput |                   index | 38138.1         | 55643.2         | 17505.2     | docs/s |  +45.90% |
|                                             Median Throughput |                   index | 37920.8         | 55406.3         | 17485.5     | docs/s |  +46.11% |
|                                                Max Throughput |                   index | 42937.3         | 58098.1         | 15160.8     | docs/s |  +35.31% |
|                                       50th percentile latency |                   index |   972.428       |   594.156       |  -378.272   |     ms |  -38.90% |
|                                       90th percentile latency |                   index |  1410.24        |  1128.94        |  -281.301   |     ms |  -19.95% |
|                                       99th percentile latency |                   index |  3633.89        |  2819.42        |  -814.469   |     ms |  -22.41% |
|                                     99.9th percentile latency |                   index |  6552.54        |  4340.73        | -2211.81    |     ms |  -33.76% |
|                                    99.99th percentile latency |                   index |  8315.43        |  5240.54        | -3074.89    |     ms |  -36.98% |
|                                      100th percentile latency |                   index |  9261           |  5770.05        | -3490.95    |     ms |  -37.70% |
|                                  50th percentile service time |                   index |   972.428       |   594.156       |  -378.272   |     ms |  -38.90% |
|                                  90th percentile service time |                   index |  1410.24        |  1128.94        |  -281.301   |     ms |  -19.95% |
|                                  99th percentile service time |                   index |  3633.89        |  2819.42        |  -814.469   |     ms |  -22.41% |
|                                99.9th percentile service time |                   index |  6552.54        |  4340.73        | -2211.81    |     ms |  -33.76% |
|                               99.99th percentile service time |                   index |  8315.43        |  5240.54        | -3074.89    |     ms |  -36.98% |
|                                 100th percentile service time |                   index |  9261           |  5770.05        | -3490.95    |     ms |  -37.70% |
|                                                    error rate |                   index |     0           |     0           |     0       |      % |    0.00% |
```

**Comment by martijnvg** [comment]
I ran the tsdb track another time. This time ensuring that the force merge task merges to a single segment, so that the disk usage increase because of enabling bloom filter for _id field is visible.

```
| Store size                      |       |     6.95962     |     7.06227     |     0.10265 |     GB |   +1.47% |
| Segment count                   |       |     1           |     1           |     0       |        |    0.00% |
| tsdb _id inverted index         |       |   455.25        |   510.262       |    55.0125  |     MB |  +12.08% |
| tsdb _id stored fields          |       |   710.812       |   690.406       |   -20.4061  |     MB |   -2.87% |
| tsdb _id total                  |       |     1.13873     |     1.17253     |     0.0338  |     GB |   +2.97% |
| Min Throughput                  | index | 36447.3         | 52903.2         | 16455.9     | docs/s |  +45.15% |
| Mean Throughput                 | index | 38637.9         | 55409.9         | 16772       | docs/s |  +43.41% |
| Median Throughput               | index | 38317.4         | 55378.6         | 17061.2     | docs/s |  +44.53% |
| Max Throughput                  | index | 42816.5         | 58012.7         | 15196.3     | docs/s |  +35.49% |
| 50th percentile latency         | index |   934.85        |   604.44        |  -330.41    |     ms |  -35.34% |
| 90th percentile latency         | index |  1379.02        |  1130.27        |  -248.755   |     ms |  -18.04% |
| 99th percentile latency         | index |  3594.48        |  2857.93        |  -736.548   |     ms |  -20.49% |
| 99.9th percentile latency       | index |  6724           |  4580.35        | -2143.65    |     ms |  -31.88% |
| 99.99th percentile latency      | index |  8461.3         |  5427.81        | -3033.49    |     ms |  -35.85% |
| 100th percentile latency        | index |  8977.22        |  7518.23        | -1458.99    |     ms |  -16.25% |
| 50th percentile service time    | index |   934.85        |   604.44        |  -330.41    |     ms |  -35.34% |
| 90th percentile service time    | index |  1379.02        |  1130.27        |  -248.755   |     ms |  -18.04% |
| 99th percentile service time    | index |  3594.48        |  2857.93        |  -736.548   |     ms |  -20.49% |
| 99.9th percentile service time  | index |  6724           |  4580.35        | -2143.65    |     ms |  -31.88% |
| 99.99th percentile service time | index |  8461.3         |  5427.81        | -3033.49    |     ms |  -35.85% |
| 100th percentile service time   | index |  8977.22        |  7518.23        | -1458.99    |     ms |  -16.25% |
| error rate                      | index |     0           |     0           |     0       |      % |    0.00% |
```

The store size in general increased with 1,47% and the inverted index for _id field increased by 12,08%.
I think the increase of disk usage for the bloom filter is worth the increase of indexing speed.

File: server/src/main/java/org/elasticsearch/index/codec/PerFieldMapperCodec.java [hunk 1]
```diff
@@ -64,10 +66,16 @@ private PostingsFormat internalGetPostingsFormatForField(String field) {
...
     }
 
-    private boolean useBloomFilter(String field) {
-        return IdFieldMapper.NAME.equals(field)
-            && mapperService.mappingLookup().isDataStreamTimestampFieldEnabled() == false
-            && IndexSettings.BLOOM_FILTER_ID_FIELD_ENABLED_SETTING.get(mapperService.getIndexSettings().getSettings());
+    boolean useBloomFilter(String field) {
+        IndexSettings indexSettings = mapperService.getIndexSettings();
+        if (mapperService.mappingLookup().isDataStreamTimestampFieldEnabled()) {
+            return indexSettings.getIndexVersionCreated().onOrAfter(Version.V_8_7_0)
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/codec/PerFieldMapperCodec.java:72]: We don't need the onOrAtfter check: PerFieldPostingsFormat records what fields use what format on disk, so old indices would keep using the normal postings format automatically.

Comment (jpountz) [reply server/src/main/java/org/elasticsearch/index/codec/PerFieldMapperCodec.java:72]: Can you also add a comment about why we're treating TSDB data streams differently?

**Review by jpountz — APPROVED** [review]
This trade-off makes sense to me: ingestion rate is one of the main things that are in the way of adoption, and this change trades a very small amount of space efficiency for much better better ingestion speed. The maintenance overhead is very small as we need to maintain bw compat for the bloom filter format anyway for regular indexes.

**Pull request merged**